File: Notes/app/src/main/java/com/serhii/apps/notes/database/impl/database/DatabaseImpl.java
/*
 * Copyright 2022. Happy coding ! :)
 * Author: Serhii Butryk
 */
package com.serhii.apps.notes.database.impl.database;

import android.content.Context;

import java.util.Map;

public abstract class DatabaseImpl {

    public static final String FLAG_SELECT = "=?";
    public static final String FLAG_ORDER = " ASC";

    protected static boolean initialized;

    public void initDbImpl(Context context) { }

    public void clearDatabaseImpl() { checkIfInitialized(); }

    public int addRecordImpl(final String data) { checkIfInitialized(); return -1; }

    public boolean deleteRecordImpl(final String id) { checkIfInitialized(); return false; }

    public boolean updateRecordImpl(final String id, final String newData) { checkIfInitialized(); return false; }

    public String getRecordImpl(final String id) { checkIfInitialized(); return  null; }

    public Map<Integer, String> getRecordsImpl() { checkIfInitialized(); return null; }

    public void closeImpl() { checkIfInitialized(); }

    public int getRecordsCountImpl() { checkIfInitialized(); return 0; }

    protected void checkIfInitialized() {
        if (!initialized) {
            throw new IllegalStateException("Database was not initialized");
        }
    }

}

File: Notes/app/src/main/java/com/serhii/apps/notes/database/impl/database/DatabaseImpl.kt
/*
 * Copyright 2022. Happy coding ! :)
 * Author: Serhii Butryk
 */
package com.serhii.apps.notes.database.impl.database

import android.content.Context

abstract class DatabaseImpl {

    protected var initialized = false

    open fun initDbImpl(context: Context) {}

    open fun clearDatabaseImpl() {
        checkIfInitialized()
    }

    open fun addRecordImpl(data: String): Int {
        checkIfInitialized()
        return -1
    }

    open fun deleteRecordImpl(id: String): Boolean {
        checkIfInitialized()
        return false
    }

    open fun updateRecordImpl(id: String, newData: String): Boolean {
        checkIfInitialized()
        return false
    }

    open fun getRecordImpl(id: String): String {
        checkIfInitialized()
        return ""
    }

    open val records: Map<Int, String>
        get() {
            checkIfInitialized()
            return emptyMap()
        }

    open fun closeImpl() {
        checkIfInitialized()
    }

    open val recordsCount: Int
        get() {
            checkIfInitialized()
            return 0
        }

    private fun checkIfInitialized() {
        check(initialized) { "Database was not initialized" }
    }

    companion object {
        const val FLAG_SELECT = "=?"
        const val FLAG_ORDER = " ASC"
    }
}